crates/vm2/src/world_diff.rs

```diff
@@ -372,102 +378,198 @@ mod tests {
+            })
+            .collect();
+        assert_eq!(actual_changes, expected_changes);
+
+        let checkpoint2 = world_diff.snapshot();
+        for (key, value) in &second_changes {
+            world_diff.write_storage(&mut NoWorld, &mut (), key.0, key.1, *value);
+        }
+        let actual_changes = world_diff
+            .get_storage_changes_after(&checkpoint2)
+            .collect::<BTreeMap<_, _>>();
+        let expected_changes = second_changes
+            .iter()
+            .map(|(key, value)| {
+                let before = first_changes
+                    .get(key)
+                    .or(initial_values.get(key).map(|slot| &slot.value))
+                    .copied()
+                    .unwrap_or_default();
+                let is_initial = initial_values
+                    .get(key)
+                    .map_or(true, |slot| slot.is_write_initial);
+                (
+                    *key,
+                    StorageChange {
+                        before,
+                        after: *value,
+                        is_initial,
+                    },
+                )
+            })
+            .collect();
+        assert_eq!(actual_changes, expected_changes);
+
+        let mut combined = first_changes
+            .into_iter()
+            .filter_map(|(key, value)| {
+                let initial = initial_values
+                    .get(&key)
+                    .copied()
+                    .unwrap_or(StorageSlot::EMPTY);
+                (initial.value != value).then_some((
+                    key,
+                    StorageChange {
+                        before: initial.value,
+                        after: value,
+                        is_initial: initial.is_write_initial,
+                    },
+                ))
+            })
+            .collect::<BTreeMap<_, _>>();
+        for (key, value) in second_changes {
+            let initial = initial_values
+                .get(&key)
+                .copied()
+                .unwrap_or(StorageSlot::EMPTY);
+            if initial.value == value {
+                combined.remove(&key);
+            } else {
+                combined.insert(
+                    key,
+                    StorageChange {
+                        before: initial.value,
+                        after: value,
+                        is_initial: initial.is_write_initial,
+                    },
+                );
+            }
+        }
+
+        assert_eq!(combined, world_diff.get_storage_changes().collect());
+    }
 
     proptest! {
         #[test]
-        fn test_storage_changes(
-            initial_values in arbitrary_storage_changes(),
+        fn storage_changes_work_as_expected(
+            initial_values in arbitrary_initial_storage(),
             first_changes in arbitrary_storage_changes(),
             second_changes in arbitrary_storage_changes(),
         ) {
-            let storage_initial_values = initial_values
```

Comment:
Moved this out of the macro for reuse, and because the Rust formatter is notoriously bad at formatting code inside macros (it mostly ignores badly formatted code).